User defined window functions blog post (#66)
apache/datafusion-site: content/blog/2025-04-04-datafusion-userdefined-window-functions.md

In this post, we'll explore:

- What window functions are and why they matter

- Understanding sliding windows

- The challenges of computing window aggregates efficiently

- How DataFusion optimizes these computations

- Alternative approaches and their trade-offs

## Understanding Window Functions in SQL 

Imagine you're analyzing sales data and want insights without losing the finer details. This is where **window functions** come into play. Unlike **GROUP BY**, which condenses data, window functions let you retain each row while performing calculations over a defined **range** —like having a moving lens over your dataset.

Picture a business tracking daily sales. They need a running total to understand cumulative revenue trends without collapsing individual transactions. SQL makes this easy:
```sql
SELECT id, value, SUM(value) OVER (ORDER BY id) AS running_total
FROM sales;
```
This helps in analytical queries where we need cumulative sums, moving averages, or ranking without losing individual records.

> ## User Defined Window Functions

Reviewer: I think DataFusion is somewhat rare in the database world for supporting user defined window functions

Perhaps we can add some background here such as "while many databases support user defined scalar and user defined aggregate functions, user defined window functions are not as common" (ideally we could find some evidence to back up that claim too)

Author: Sure !, i would start working on that

Writing a user defined window function is slightly more complex than an aggregate function due
to the variety of ways that window functions are called. I recommend reviewing the

> [online documentation](https://datafusion.apache.org/library-user-guide/adding-udfs.html)

Reviewer: It might help to link to the subsection https://datafusion.apache.org/library-user-guide/adding-udfs.html#registering-a-window-udf

Author: Done 👍

for a description of which functions need to be implemented. The details of how to implement

> these generally follow the same patterns as described above for aggregate functions.

Reviewer: I don't see any description above for aggregate functions 🤔

Author: I had updated the link you can now review it ! @alamb

## Understaing Sliding Window 

Sliding windows define a **moving range** of data over which aggregations are computed. Unlike simple cumulative functions, these windows are dynamically updated as new data arrives.

For instance, if we want a 7-day moving average of sales:

```sql
SELECT date, sales, 
       AVG(sales) OVER (ORDER BY date ROWS BETWEEN 6 PRECEDING AND CURRENT ROW) AS moving_avg
FROM sales;
```
Here, each row’s result is computed based on the last 7 days, making it computationally intensive as data grows.

## Why Computing Sliding Windows Is Hard

> Imagine you’re at a café, and the barista is preparing coffee orders. If they made each cup from scratch without using pre-prepared ingredients, the process would be painfully slow. This is exactly the problem with naïve sliding window computations.

Reviewer: this is a pretty cool analogy

Author: thank you

Computing sliding windows efficiently is tricky because:

- **High Computation Costs:** Just like making coffee from scratch for each customer, recalculating aggregates for every row is expensive.

- **Data Shuffling:** In large distributed systems, data must often be shuffled between nodes, causing delays—like passing orders between multiple baristas who don’t communicate efficiently.

- **State Management:** Keeping track of past computations is like remembering previous orders without writing them down—error-prone and inefficient.

Many traditional query engines struggle to optimize these computations effectively, leading to sluggish performance.

> ## How DataFusion Making it fast

Reviewer: ```suggestion
## How DataFusion Evaluates Window Functions Quickly
```

In the world of big data, every millisecond counts. Imagine you’re analyzing stock market data, tracking sensor readings from millions of IoT devices, or crunching through massive customer logs—speed matters. This is where [DataFusion](https://datafusion.apache.org/) shines, making window function computations blazing fast. Let’s break down how it achieves this remarkable performance.

DataFusion now supports [user-defined window aggregates (UDWAs)](https://datafusion.apache.org/library-user-guide/adding-udfs.html), meaning you can bring your own aggregation logic and use it within a window function.

```sql
let my_udwa = create_my_custom_udwa();
ctx.register_udaf("my_moving_avg", my_udwa);

// Then use in SQL:
SELECT
  user_id,
  my_moving_avg(score) OVER (
    PARTITION BY user_id
    ORDER BY game_time
    ROWS BETWEEN 2 PRECEDING AND CURRENT ROW
  ) AS moving_score
FROM leaderboard;
```

> This gives you full flexibility to build **domain-specific logic** that plugs seamlessly into DataFusion’s engine — all without sacrificing performance.

Reviewer: It would be great to expand this section a little more with what a user defined aggregate looks like -- perhaps copying soem/all of the example from https://datafusion.apache.org/library-user-guide/adding-udfs.html#registering-a-window-udf

Author: 👍

Author: done that

## Performance Gains

> To demonstrate efficiency, we benchmarked a 1-million row dataset with a sliding window aggregate.

Reviewer: If we are going to report numbers like this I think it is important to be clear what query was run on what setup (so they can be reproduced if desired)

Author: what can be ideal approach ? . it would be great if anyone wants to help with this specific section @alamb  @Dandandan

Reviewer: I suggest we avoid comparisons with other systems and simply focus on DataFusion's own performance.

Author: so do i need to remove this section ?

Author: okay , i also feel it quite so

```
+--------------------------+----------------------+
| Engine                   | Query Execution Time |
+--------------------------+----------------------+
| PostgeSQL                |   1.2s               |
| Spark                    |   0.9s               |
| DataFusion               |   0.45s              |
+--------------------------+----------------------+

```
DataFusion outperforms traditional SQL engines by leveraging [Apache Arrow](https://arrow.apache.org/) optimizations, making it a great choice for analytical workloads .

> Note: The reference has been taken from [@andygrove]'s blog . [see](https://andygrove.io/2019/04/datafusion-0.13.0-benchmarks/)

Reviewer 2: I don't see these numbers?

Author: ![image](https://github.com/user-attachments/assets/e66b4e47-8864-4b73-bd1c-f481a2e93085)

Yes @Dandandan  you are right !. Actually i took the reference from here but if  had any better suggestions or stats i would welcome you to please help with that . once again thank you for your consideration

## Final Thoughts and Recommendations 

> With the addition of sliding window support and user-defined aggregates, DataFusion continues its march toward being a high-performance analytical engine that balances power, extensibility, and speed.

Reviewer: Given window functions were added a while ago we may want to emphasize that DataFUsion has had window functions since XXX (look up when they were added)

Author: I am not able to find the dates when this got released it would be great help

> Window functions may be common in SQL, but *efficient and extensibl*e window engines are rare — and now DataFusion is one of them.

Reviewer 2: ```suggestion
Window functions may be common in SQL, but *efficient and extensible* window engines are rare — and now DataFusion is one of them.
```

> Window functions may be common in SQL, but *efficient and extensible* window engines are rare — and now DataFusion is one of them.

Reviewer: this might be a good sentence to move up

Author: Sure !

{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

## Introduction
Window functions are a powerful feature in SQL, allowing for complex analytical computations over a subset of data. However, efficiently implementing them, especially sliding windows, can be quite challenging. With DataFusion's recent support for user-defined window functions , developers now have more flexibility and performance improvements at their disposal.

In this post, we'll explore:

> - What window functions are and why they matter

Author: will do that

## User Defined Window Functions
Built-in window functions serve many use cases, but sometimes custom logic is needed—for example:

- Calculating moving averages with complex conditions

- Implementing a custom ranking strategy

- Tracking non-standard cumulative logic

Thus, **User-Defined Window Functions (UDWFs)** allow developers to define their own behavior using a combination of SQL and *bit of logic*.

Writing a user defined window function is slightly more complex than an aggregate function due
to the variety of ways that window functions are called. I recommend reviewing the
[online documentation](https://datafusion.apache.org/library-user-guide/adding-udfs.html#registering-a-window-udf)
for a description of which functions need to be implemented. 

## Understaing Sliding Window 

Sliding windows define a **moving range** of data over which aggregations are computed. Unlike simple cumulative functions, these windows are dynamically updated as new data arrives.

For instance, if we want a 7-day moving average of sales:

```sql
SELECT date, sales, 
       AVG(sales) OVER (ORDER BY date ROWS BETWEEN 6 PRECEDING AND CURRENT ROW) AS moving_avg
FROM sales;
```

> Here, each row’s result is computed based on the last 7 days, making it computationally intensive as data grows.

Reviewer: I am not sure I would say a 7 day window is computationally expensive -- if you changed the example to be `UNBOUNDED PRECEDING` you could make the point that the window size kept growing. Or 365 days maybe would make the point better?

## Why Computing Sliding Windows Is Hard

Imagine you’re at a café, and the barista is preparing coffee orders. If they made each cup from scratch without using pre-prepared ingredients, the process would be painfully slow. This is exactly the problem with naïve sliding window computations.

Computing sliding windows efficiently is tricky because:

- **High Computation Costs:** Just like making coffee from scratch for each customer, recalculating aggregates for every row is expensive.

- **Data Shuffling:** In large distributed systems, data must often be shuffled between nodes, causing delays—like passing orders between multiple baristas who don’t communicate efficiently.

- **State Management:** Keeping track of past computations is like remembering previous orders without writing them down—error-prone and inefficient.

Many traditional query engines struggle to optimize these computations effectively, leading to sluggish performance.

## How DataFusion Evaluates Window Functions Quickly

> In the world of big data, every millisecond counts. Imagine you’re analyzing stock market data, tracking sensor readings from millions of IoT devices, or crunching through massive customer logs—speed matters. This is where [DataFusion](https://datafusion.apache.org/) shines, making window function computations blazing fast. Let’s break down how it achieves this remarkable performance.

Reviewer: This feels a little over dramatic -- perhaps we can tone down the wording some

In general I don't think DataFusion's window function performance is significantly better (or worse) than other engines . I think the main difference is how extensible it is. I'll try and comment more inline

DataFusion now supports [user-defined window aggregates (UDWAs)](https://datafusion.apache.org/library-user-guide/adding-udfs.html), meaning you can bring your own aggregation logic and use it within a window function.

For example, we will declare a user defined window function that computes a moving average.
```sql
use datafusion::arrow::{array::{ArrayRef, Float64Array, AsArray}, datatypes::Float64Type};
use datafusion::logical_expr::{PartitionEvaluator};
use datafusion::common::ScalarValue;
use datafusion::error::Result;
/// This implements the lowest level evaluation for a window function
///
/// It handles calculating the value of the window function for each
/// distinct values of `PARTITION BY`
#[derive(Clone, Debug)]
struct MyPartitionEvaluator {}

impl MyPartitionEvaluator {
    fn new() -> Self {
        Self {}
    }
}

> /// Different evaluation methods are called depending on the various
> /// settings of WindowUDF. This example uses the simplest and most
> /// general, `evaluate`. See `PartitionEvaluator` for the other more
> /// advanced uses.
> impl PartitionEvaluator for MyPartitionEvaluator {

Reviewer: In this blog form, it might be easier to read if you broke the comments out of the text so something lik e

```suggestion
```(end text)
Different evaluation methods are called depending on the various
settings of WindowUDF. In the first example, we  use the simplest and most
general, `evaluate`. we will seee how to use `PartitionEvaluator` for the other more
 advanced uses later in the article.
 
 ```rust
impl PartitionEvaluator for MyPartitionEvaluator {
```

    /// considered for the calculation.
    ///
    /// Note this is the SLOWEST, but simplest, way to evaluate a
    /// window function. It is much faster to implement
    /// evaluate_all or evaluate_all_with_rank, if possible
    fn evaluate(
        &mut self,
        values: &[ArrayRef],
        range: &std::ops::Range<usize>,
    ) -> Result<ScalarValue> {
        // Again, the input argument is an array of floating
        // point numbers to calculate a moving average
        let arr: &Float64Array = values[0].as_ref().as_primitive::<Float64Type>();

        let range_len = range.end - range.start;

        // our smoothing function will average all the values in the
        let output = if range_len > 0 {
            let sum: f64 = arr.values().iter().skip(range.start).take(range_len).sum();
            Some(sum / range_len as f64)
        } else {
            None
        };

        Ok(ScalarValue::Float64(output))
    }
}

/// Create a `PartitionEvaluator` to evaluate this function on a new
/// partition.
fn make_partition_evaluator() -> Result<Box<dyn PartitionEvaluator>> {
    Ok(Box::new(MyPartitionEvaluator::new()))
}
```
### Registering a Window UDF

> To register a Window UDF, you need to wrap the function implementation in a `WindowUDF` struct and then register it with the `SessionContext`. DataFusion provides the `create_udwf` helper functions to make this easier. There is a lower level API with more functionality but is more complex, that is documented in [advanced_udwf.rs](https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/advanced_udwf.rs).

Reviewer: It would be nice if you could add links to the doc for `WindowUDF`, `SessionContext`, etc here

Author: got it

---
layout: post
title: User defined Window Functions in DataFusion 
date: 2025-04-04
author: Aditya Singh Rathore
categories: [tutorial]
---

<!--
{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

> ## Introduction

Author: will do it now !
